core/src/main/java/org/elasticsearch/index/query/FilteredQueryParser.java

```diff
@@ -83,44 +78,15 @@ public Query parse(QueryParseContext parseContext) throws IOException, QueryPars
             }
         }
 
-        // parsed internally, but returned null during parsing...
-        if (query == null) {
-            return null;
-        }
-
-        if (filter == null) {
-            if (!filterFound) {
-                // we allow for null filter, so it makes compositions on the client side to be simpler
-                return query;
-            } else {
-                // even if the filter is not found, and its null, we should simply ignore it, and go
-                // by the query
-                return query;
-            }
-        }
-        if (Queries.isConstantMatchAllQuery(filter)) {
-            // this is an instance of match all filter, just execute the query
-            return query;
-        }
-
-        // if its a match_all query, use constant_score
-        if (Queries.isConstantMatchAllQuery(query)) {
-            Query q = new ConstantScoreQuery(filter);
-            q.setBoost(boost);
-            return q;
-        }
-
-        BooleanQuery filteredQuery = Queries.filtered(query, filter);
-
-        filteredQuery.setBoost(boost);
-        if (queryName != null) {
-            parseContext.addNamedQuery(queryName, filteredQuery);
-        }
-        return filteredQuery;
```

Comment:
Sure - waiting for #11915 makes sense here.

About null vs. match_all: Before we would return match_all in the parser if the query part of the filter was missing in the json definition. We would return null if the query part was present but parsed to null. 

Going through the REST API this is still true as the query part gets initialised with a MatchAllQueryBuilder.

Maybe the right thing to do to carry that semantic over to the Java API would be to also initialize query to MatchAllQueryBuilder there but let the user set it to null if they want to.